diffrax/_solver/kl.py

```diff
@@ -0,0 +1,299 @@
+import operator
+from typing import Optional, Tuple, Union
+
+import equinox as eqx
+import jax
+import jax.tree_util as jtu
+import lineax as lx
+from jax import numpy as jnp
+from jaxtyping import PyTree
+
+from .._custom_types import (
+    Args,
+    BoolScalarLike,
+    Control,
+    DenseInfo,
+    RealScalarLike,
+    VF,
+    Y,
+)
+from .._heuristics import is_sde
+from .._solution import RESULTS
+from .._term import (
+    AbstractTerm,
+    ControlTerm,
+    MultiTerm,
+    ODETerm,
+    WeaklyDiagonalControlTerm,
+)
+from .base import (
+    _SolverState,
+    AbstractSolver,
+    AbstractWrappedSolver,
+)
+
+
+_DiffusionTerm = Union[ControlTerm, WeaklyDiagonalControlTerm]
```

Comment:
I don't think we want to specialise on specifically a weakly-diagonal kind of control.

Rather, I think any kind of control should be admissible! Then we use Lineax to do the linear solve against whatever kind of operator is returned.

You know, maybe what we should really be doing is implementing #370! That is to say, allow the `vector_field` of `diffrax.ControlTerm(vector_field, ...)` to instead return a `lx.AbstractLinearOperator`. That would then make things exceptionally easy over here.

WDYT?

(I'll hold off on doing the rest of the review yet as I can see this may change things.)

Reply:
I think that makes sense, that could reduce the parsing/if elsing complexity I have by a fair amount. I will make a PR for #370 then incorporate those changes into this and we can revisit this PR at that point.